### PR TITLE
Faster package/group/org extras using jsonb storage

### LINKED ISSUE #8273: Faster extras
Let's migrate package and group extras into a JSON field on the package/group table for faster loading and updating

### PR DESCRIPTION
**Pull request by wardi**
Fixes #8273 

### Proposed fixes:
drop package and group extras tables and replace them with jsonb storage in the group/package tables.

Maintain compatibility with API (continue to present and accept `extras` as a list of `{"key":…, "value":…}` dicts)

At the model level there are some developer-visible changes:
- `extras` continues to be a dict-like object but this is now the actual value stored in the DB
- `_extras` is removed
- `extras_list` is removed
- new model objects start with `extras = None`, before `extras` was always a dict-like object (this could be fixed with a `mapped_column` definition once we upgrade to sqlalchemy 2.0)

```mermaid
---
config:
    xyChart:
        height: 300
---
xychart-beta horizontal
    title "package_dict_save (50 extras)"
    x-axis ["old", "new"]
    y-axis "Speed (calls/s)" 0 --> 40
    bar [24.0, 30.6]
```

Also a *very* large table and index space savings in the database depending on the number of extras used.

### Features:

- [ ] includes tests covering changes
- [ ] includes updated documentation
- [ ] includes user-visible changes
- [ ] includes API changes
- [ ] includes bugfix for possible backport

### CODE REVIEW

File: ckan/migration/versions/106_12d64978ab44_faster_extras.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+            select json_object(
+                array_agg(package_extra.key), array_agg(package_extra.value))
+            from package_extra
+            where package.id = package_extra.package_id
+            and package_extra.state = 'active'
+        )
+        """
+    )
+    op.drop_index('idx_group_extra_group_id', table_name='group_extra')
+    op.drop_table('group_extra')
```
[the remark this answers is not in the document]
Comment (wardi) [reply ckan/migration/versions/106_12d64978ab44_faster_extras.py:67]: All the revision tables can be dropped but was planning a separate PR for that (they were left to let people migrate revisions to activities after upgrading)

[the remark this answers is not in the document]
Comment (tino097) [reply ckan/migration/versions/106_12d64978ab44_faster_extras.py:67]: @wardi about the schemas, trying to understand what impact will have the `default_extras_schema` or also that would be obsolete and removed in next releases ?

[the remark this answers is not in the document]
Comment (wardi) [reply ckan/migration/versions/106_12d64978ab44_faster_extras.py:67]: The APIs aren't affected by this change so the extras schemas will stay

**Review by tino097 — APPROVED** [review]
Good one @wardi

**Comment by amercader** [comment]
@wardi @tino097 I pulled this out of backporting because it might have a bigger impact that we think and I think it's worth discussing.

I noticed that the ckanext-dcat tests started [failing](https://github.com/ckan/ckanext-dcat/actions/runs/9775939668/job/26987437041?pr=281) after merging this, basically because of usage of `model.PackageExtra` in the harvesters:

https://github.com/ckan/ckanext-dcat/blob/83495ba99cba17398ba8feb1bc0da486f3798584/ckanext/dcat/harvesters/base.py#L164

Also the [ckanext-harvest ](https://github.com/ckan/ckanext-harvest/actions/runs/9776155745/job/26988748641) ones (because of explicitly deleting things from the package_extra table)

Usage of `PackageExtra` seems to be quite widespread across extensions:

https://github.com/search?q=PackageExtra+language%3APython++path%3A%2F%5Eckanext%5C%2F%2F&type=code&p=2

Granted, a lot of these are forks or really old versions but some of the usage seems legitimate, and if we want to drop these models we should provide a clear upgrade path and/or deprecation notices in changelog (which right now doesn't include any mention of the dropped tables and models)

**Comment by wardi** [comment]
@amercader here's a sentence about the model removal https://github.com/ckan/ckan/pull/8318 . For a deprecation notice for 2.11 where would I put that? I could add a changes file that includes the warning so that it gets pulled into the changelog, should we also try to detect when an extension imports the model and issue a warning in the logs?

**Pull request merged**